common/src/main/java/io/netty/util/internal/StringUtil.java

```diff
@@ -314,6 +322,57 @@ public static String simpleClassName(Class<?> clazz) {
         }
     }
 
+    /**
+     * Escapes the specified value, if necessary according to
+     * <a href="https://tools.ietf.org/html/rfc4180#section-2">RFC-4180</a>.
+     *
+     * @param value The value which will be escaped according to
+     *              <a href="https://tools.ietf.org/html/rfc4180#section-2">RFC-4180</a>
+     * @return {@link CharSequence} the escaped value if nesessary, or the value unchanged
+     */
+    public static CharSequence escapeCsv(CharSequence value) {
+        int length = checkNotNull(value, "value").length();
+        if (length == 0) {
+            return value;
+        }
+        int last = length - 1;
+        boolean quoted = isDoubleQuote(value.charAt(0)) && isDoubleQuote(value.charAt(last)) && length != 1;
+        boolean foundSpecialCharacter = false;
+        boolean foundEscapeCharacter = false;
+        StringBuilder escaped = new StringBuilder(length + DEFAULT_NR_OF_DOUBLE_QUOTES).append(DOUBLE_QUOTE);
+        for (int i = 0; i < length; i++) {
+            char current = value.charAt(i);
+            switch (current) {
+                case DOUBLE_QUOTE:
+                    if (i == 0 || i == last) {
+                        if (!quoted) {
+                            escaped.append(DOUBLE_QUOTE);
+                        } else {
+                            continue;
+                        }
+                    } else {
+                        boolean isNextCharDoubleQuote = isDoubleQuote(value.charAt(i + 1));
+                        if (!isDoubleQuote(value.charAt(i - 1)) &&
+                                (!isNextCharDoubleQuote || isNextCharDoubleQuote && i + 1 == last)) {
+                            escaped.append(DOUBLE_QUOTE);
+                        }
+                    }
+                    foundEscapeCharacter = true;
+                case LINE_FEED:
+                case CARRIAGE_RETURN:
+                case COMMA:
+                    foundSpecialCharacter = true;
+            }
+            escaped.append(current);
+        }
+        return foundEscapeCharacter || (foundSpecialCharacter && !quoted) ?
```

Comment:
Sorry for going silent, have another task here at work which I need to give
higher priority. Will circle back as soon as I can.

tisdag 27 januari 2015 skrev Scott Mitchell <notifications@github.com>:

> In common/src/main/java/io/netty/util/internal/StringUtil.java
> <https://github.com/netty/netty/pull/3271#discussion_r23604716>:
>
> > +                    } else {
> > +                        boolean isNextCharDoubleQuote = isDoubleQuote(value.charAt(i + 1));
> > +                        if (!isDoubleQuote(value.charAt(i - 1)) &&
> > +                                (!isNextCharDoubleQuote || isNextCharDoubleQuote && i + 1 == last)) {
> > +                            escaped.append(DOUBLE_QUOTE);
> > +                        }
> > +                    }
> > +                    foundEscapeCharacter = true;
> > +                case LINE_FEED:
> > +                case CARRIAGE_RETURN:
> > +                case COMMA:
> > +                    foundSpecialCharacter = true;
> > +            }
> > +            escaped.append(current);
> > +        }
> > +        return foundEscapeCharacter || (foundSpecialCharacter && !quoted) ?
>
>  but we do check if the character before and after the current ", and
> don't add it unless it is required
>
> Understood I think the escaping " characters looks good. I am still a bit
> uncertain about the condition which wraps the entire value in quotes. The
> foundEscapeCharacter is set unconditionally if a " is found.
> foundEscapeCharacter is used here to determine if the "quoted" version of
> the input should be used. Using this logic wouldn't it be possible to find
> a string with a " that is properly escaped and quoted, but then this code
> will still wrap the string in additional " characters? Is this line the
> same in the un-pushed code?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/netty/netty/pull/3271/files#r23604716>.
>